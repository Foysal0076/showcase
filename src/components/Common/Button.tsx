import { cva } from 'class-variance-authority'
import clsx from 'clsx'

import { Spinner } from '@/components/Common/Spinner'

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: 'primary' | 'secondary' | 'tertiary' | 'danger' | 'success'
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
  fullWidth?: boolean
  loading?: boolean
}

const ButtonVariants = cva(
  'rounded-sm transition-all duration-200 text-nowrap focus:outline-none font-medium',
  {
    variants: {
      variant: {
        primary:
          'bg-primary-500 text-white hover:bg-primary-600 focus:ring-4 focus:ring-primary-100 text-neutral-0 focus:outline-primary-200 disabled:active:transform-none active:scale-[.9] disabled:text-neutral-200 disabled:bg-neutral-50 disabled:cursor-not-allowed disabled:bg-neutral-50',
        secondary:
          'bg-none text-neutral-900 hover:bg-primary-50 focus:ring-primary-100 border-2 border-primary-500 text-primary-500 focus:ring-4 focus:ring-primary-100 disabled:active:transform-none active:scale-[.9] disabled:cursor-not-allowed disabled:text-neutral-200  disabled:border-neutral-50 disabled:hover:bg-neutral-50 dark:text-primary-200 dark:border-primary-200 dark:hover:bg-neutral-800 dark:disabled:border-neutral-200 dark:disabled:text-neutral-200',
        tertiary:
          'bg-none hover:bg-primary-50 focus:ring-4 focus:ring-primary-100 focus:text-primary-500 disabled:active:transform-none active:scale-[.9] disabled:cursor-not-allowed disabled:text-neutral-200 disabled:hover:bg-neutral-50 dark:hover:bg-neutral-800 dark:focus:text-primary-100',
        danger:
          'bg-danger-500 hover:bg-danger-600 text-neutral-0 focus:ring-4 focus:ring-danger-100 disabled:active:transform-none active:scale-[.9] disabled:text-neutral-200 disabled:bg-neutral-50 disabled:cursor-not-allowed',
        success:
          'bg-success-500 hover:bg-success-600 text-neutral-0 focus:ring-4 focus:ring-success-100 disabled:active:transform-none active:scale-[.9] disabled:text-neutral-200 disabled:bg-neutral-50 disabled:cursor-not-allowed',
      },
      size: {
        xs: 'h-[1.75rem] text-xs px-4',
        sm: 'h-[2rem] text-sm px-6',
        md: 'h-[2.5rem] text-md px-8',
        lg: 'h-[3rem] text-md px-11',
        xl: 'h-[3.5rem] text-lg px-11',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

const Button = ({
  variant = 'primary',
  size = 'md',
  loading = false,
  className = '',
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      className={clsx(
        `relative select-none`,
        `${ButtonVariants({ variant, size })}`,
        `${className}`
      )}
      disabled={loading}
      {...buttonProps}>
      {loading && (
        <div className='absolute inset-0 flex items-center justify-center'>
          <Spinner variant='neutral' />
        </div>
      )}
      <div
        className={clsx(`visible`, {
          invisible: loading,
          'opacity-0': loading,
        })}>
        {children}
      </div>
    </button>
  )
}

export default Button
