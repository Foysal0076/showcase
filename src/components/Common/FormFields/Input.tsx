import clsx from 'clsx'
import React, { forwardRef } from 'react'

type TextFieldParams = React.InputHTMLAttributes<HTMLInputElement> & {
  label?: string
  helperText?: string
  layout?: 'horizontal' | 'vertical'
  error?: string
  startAdornment?: StartAdornment
  endAdornment?: EndAdornment
}

type StartAdornment = {
  adornment: React.ReactNode
  onClick: () => void
  className?: string
}

type EndAdornment = {
  adornment: React.ReactNode
  onClick: () => void
  className?: string
}

export const Input = forwardRef<HTMLInputElement, TextFieldParams>(
  function CustomInput(
    {
      label,
      error,
      layout = 'vertical',
      className = '',
      helperText,
      startAdornment,
      endAdornment,
      ...inputParams
    },
    ref
  ) {
    return (
      <div
        className={clsx('custom-input group flex', {
          'flex-col gap-1': layout === 'vertical',
        })}>
        {label && (
          <label
            htmlFor={inputParams.id as string}
            className={clsx(
              'font-semibold',
              {
                'text-neutral-600 group-focus-within:text-primary-500 dark:text-neutral-50 dark:group-focus-within:text-primary-300':
                  !error,
              },
              { 'text-danger-500 group-focus-within:text-danger-500': error },
              {
                'after:ml-0.5 after:mr-1 after:text-danger-500 after:content-["*"]':
                  inputParams.required,
              },
              {
                'mt-2 ': layout === 'horizontal',
              }
            )}>
            {label}
            {layout === 'horizontal' && <span>:</span>}
          </label>
        )}
        <div className='relative flex flex-col'>
          <input
            ref={ref}
            {...inputParams}
            className={clsx(
              `rounded-sm border-none px-4 text-neutral-600 outline-none ring-1 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-neutral-800 dark:text-neutral-50 dark:disabled:bg-neutral-400 ${className}`,
              {
                'ring-neutral-50 focus:ring-primary-500 dark:focus:ring-primary-300':
                  !error,
                'ring-danger-500 focus:ring-danger-500': error,
                'pr-12': endAdornment,
                'pl-12': startAdornment,
              }
            )}
          />
          {helperText && (
            <span className='helper-text text-xs font-normal text-neutral-400 md:text-sm'>
              {`${helperText}`}
            </span>
          )}
          {error && (
            <div
              className={clsx(
                'absolute -bottom-0 left-0 ',
                { 'translate-y-[80%]': helperText },
                { 'translate-y-[90%]': !helperText }
              )}>
              <span className='-mt-1 text-xs text-danger-500 dark:text-danger-500'>
                {error}
              </span>
            </div>
          )}
          {startAdornment && (
            <button
              onClick={startAdornment.onClick}
              className={clsx(
                'absolute left-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                `${startAdornment.className ? startAdornment.className : ''}`
              )}
              aria-label='start-adornment btn'>
              {startAdornment.adornment}
            </button>
          )}
          {endAdornment && (
            <button
              onClick={endAdornment.onClick}
              className={clsx(
                'absolute right-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                `${endAdornment.className ? endAdornment.className : ''}`
              )}
              aria-label='end-adornment btn'>
              {endAdornment.adornment}
            </button>
          )}
        </div>
      </div>
    )
  }
)
