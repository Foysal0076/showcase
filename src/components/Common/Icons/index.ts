export { ToggleDarkIcon } from './ToggleDarkIcon'
export { ToggleLightIcon } from './ToggleLightIcon'
export { EyeHideIcon } from './EyeHideIcon'
export { EyeShowIcon } from './EyeShowIcon'
