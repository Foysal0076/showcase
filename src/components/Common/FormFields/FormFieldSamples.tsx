'use client'

import { Input } from '@/components/Common/FormFields/Input'
import { EyeHideIcon, EyeShowIcon } from '@/components/Common/Icons'

const FormFieldSamples = () => {
  return (
    <div className='flex flex-col gap-10'>
      <div className='flex flex-col gap-6 rounded border border-primary-100 p-4'>
        <h2>Input Field</h2>
        <Input label='Text Input' placeholder='Placeholder text' />
        <Input
          label='Text Input Disabled'
          placeholder='Placeholder text'
          disabled
        />
        <Input
          label='Required Text Input'
          required
          helperText='This is label helper text'
          placeholder='Placeholder text'
        />
        <Input
          label='Required Error Text Input'
          required
          error='This is an error message'
          placeholder='Placeholder text'
        />
        <Input
          label='Required Error Text Input Horizontal'
          required
          error='This is an error message'
          placeholder='Placeholder text'
          layout='horizontal'
          helperText='This is label helper text'
        />
        <Input
          type='text'
          label='Required Error Text Input'
          required
          placeholder='********'
          startAdornment={{
            adornment: <EyeHideIcon />,
            onClick: () => {},
            className:
              'text-neutral-500 hover:text-neutral-700 transition-colors duration-300',
          }}
        />
        <Input
          type='password'
          label='Required Error Text Input'
          required
          placeholder='********'
          endAdornment={{
            adornment: <EyeShowIcon />,
            onClick: () => {},
            className:
              'text-neutral-500 hover:text-neutral-700 transition-colors duration-300',
          }}
        />
      </div>
    </div>
  )
}

export default FormFieldSamples
