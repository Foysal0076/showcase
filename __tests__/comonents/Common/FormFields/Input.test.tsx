import { render, screen, fireEvent } from '@testing-library/react'
import { Input } from '@/components/Common/FormFields'

describe('Input', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Input
        label='Test Label'
        helperText='Test Helper Text'
        error='Test Error'
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders correctly', () => {
    render(<Input />)
    const inputElement = screen.getByRole('textbox')
    expect(inputElement).toBeInTheDocument()
  })

  it('displays label correctly', () => {
    render(<Input label='Test Label' />)
    const labelElement = screen.getByText('Test Label')
    expect(labelElement).toBeInTheDocument()
  })

  it('displays helper text correctly', () => {
    render(<Input helperText='Test Helper Text' />)
    const helperTextElement = screen.getByText('Test Helper Text')
    expect(helperTextElement).toBeInTheDocument()
  })

  it('displays error message correctly', () => {
    render(<Input error='Test Error' />)
    const errorElement = screen.getByText('Test Error')
    expect(errorElement).toBeInTheDocument()
  })

  it('handles start adornment click correctly', () => {
    const handleClick = jest.fn()
    render(
      <Input startAdornment={{ adornment: 'Start', onClick: handleClick }} />
    )
    const startAdornmentElement = screen.getByLabelText('start-adornment btn')
    fireEvent.click(startAdornmentElement)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('handles end adornment click correctly', () => {
    const handleClick = jest.fn()
    render(<Input endAdornment={{ adornment: 'End', onClick: handleClick }} />)
    const endAdornmentElement = screen.getByLabelText('end-adornment btn')
    fireEvent.click(endAdornmentElement)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })
})
