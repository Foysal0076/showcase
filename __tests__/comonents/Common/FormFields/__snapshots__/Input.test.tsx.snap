// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<DocumentFragment>
  <div
    class="custom-input group flex flex-col gap-1"
  >
    <label
      class="font-semibold text-danger-500 group-focus-within:text-danger-500"
    >
      Test Label
    </label>
    <div
      class="relative flex flex-col"
    >
      <input
        class="rounded-sm border-none px-4 text-neutral-600 outline-none ring-1 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-neutral-800 dark:text-neutral-50 dark:disabled:bg-neutral-400  ring-danger-500 focus:ring-danger-500"
      />
      <span
        class="helper-text text-xs font-normal text-neutral-400 md:text-sm"
      >
        Test Helper Text
      </span>
      <div
        class="absolute -bottom-0 left-0  translate-y-[80%]"
      >
        <span
          class="-mt-1 text-xs text-danger-500 dark:text-danger-500"
        >
          Test Error
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
